### PR TITLE
[react]: React.createElement

### PR DESCRIPTION
**Pull request by xiaojingzhao**

### CODE REVIEW

File: react/createElement.md [hunk 1]
```diff
@@ -0,0 +1,370 @@
+# 2019-12-15
```
Comment (xiaojingzhao) [inline react/createElement.md]: title should be createElement 分析

File: react/createElement.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
```
Comment (xiaojingzhao) [inline react/createElement.md:2]: Add section **abstract**

File: react/createElement.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+createElement.apply(this, arguements); // 等价于 createElement(type, props, children)
+```
+
+这个函数主要做以下事情
+
+### 2.1. 初始化 element 需要的参数
+
+```js
+var props = {};
+var key = null;
```
Comment (xiaojingzhao) [inline react/createElement.md]: what‘s the type and function of these variables？

File: react/createElement.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+### 2.2. 解析 config
+
+config 是 React.createElement 的第二个参数 props
+
+```js
+// config
+// 如果存在class或者一些其他属性，也会被记录在 config中
+{
+  __source: {
+    fileName: _jsxFileName,
```
Comment (xiaojingzhao) [inline react/createElement.md]: give an explanation for this config example？

File: react/createElement.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+- 拿到 config 中的 key 和 ref，在这个例子中没有 key 和 ref
+- 根据 config 生成一份 props，这个 props 里不包括 `key`, `ref`, `__self`, `__source`
+
+#### 2.2.1 存在 key, ref 和其他参数的情况
+
+```js
+ReactDOM.render(<><App key="app" newProps="test" ref={ele => console.log(ref)} /></>, rootElement;
+```
+
+传入的 config Object
```
Comment (xiaojingzhao) [inline react/createElement.md]: what’s the difference between the config mentioned above and this one？

File: react/createElement.md [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+生成的 props，key 和 ref 并不在 props 上。
+
+```js
+{
+  newProps: "test";
+}
+```
+
+### 2.3. 拿到 children
```
Comment (xiaojingzhao) [inline react/createElement.md]: ”拿到children“ => ”设置好 children”？

File: react/createElement.md [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+  if (Object.freeze) {
+    Object.freeze(childArray);
+  }
+}
+props.children = childArray;
+```
+
+### 2.4. defaultProps
+
+defaultProps 只会给属性是 undefined 的情况的 key 设置值。是 null 或者是""都不会覆盖。
```
Comment (xiaojingzhao) [inline react/createElement.md]: 。=> ,

**Review by xiaojingzhao — COMMENTED** [review]
1. 统一副标题
2. 多解释需要解释的，简单的罗列没有阅读价值，可以这样想想：这段代码为什么需要这么写？如果让我来写我会怎么写？

File: react/createElement.md [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+如果 type 是 object，type 不能是 null。 type.\$\$typeof 可以是以下
+
+```js
+type.$$typeof === REACT_LAZY_TYPE ||
+type.$$typeof === REACT_MEMO_TYPE || // React.memo
+type.$$typeof === REACT_PROVIDER_TYPE || // React.createContext().Provider
+type.$$typeof === REACT_CONTEXT_TYPE || // React.createContext().Consumer
+type.$$typeof === REACT_FORWARD_REF_TYPE || // React.createContext().
+type.$$typeof === REACT_FUNDAMENTAL_TYPE || // React.forwardRef
+  type.$$typeof === REACT_RESPONDER_TYPE ||
```
Comment (ShiKaiWi) [inline react/createElement.md]: seems wrong indent

Comment (ShiKaiWi) [inline react/createElement.md]: seems wrong intent

**Pull request merged**